yari-py/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+use pyo3::create_exception;
+use pyo3::exceptions::PyException;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use pyo3::types::PyString;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::str::FromStr;
+use yari_sys::Context as YARIContext;
+use yari_sys::ContextBuilder;
+use yari_sys::Module;
+use yari_sys::YrValue;
+
+// Exception type for yari python module.
+create_exception!(yari, YariError, PyException);
+
+/// Python `Context` wrapper.
+#[pyclass(unsendable)]
+struct Context {
+    inner: YARIContext,
+}
+
+/// Helper to convert `YrValue` to a `PyObject`.
+///
+/// Return values are modeled similar to Python built-in `eval` function. This function converts
+/// values to the Python native types. For example `Vec<YrValue>` is converted to `list[YrValue]`.
+/// Same applies to structures and dictionaries.
+fn yr_value_to_py_object(yr_value: &YrValue) -> PyObject {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    match yr_value {
+        YrValue::Integer(i) => i.into_py(py),
+        YrValue::Float(f) => f.into_py(py),
+        YrValue::String(s) => s.into_py(py),
+        YrValue::Dictionary(d) => d
+            .iter()
+            .map(|(k, v)| (k, yr_value_to_py_object(v)))
+            .collect::<HashMap<_, _>>()
+            .into_py(py),
+        YrValue::Array(a) => a
+            .iter()
+            .map(yr_value_to_py_object)
+            .collect::<Vec<_>>()
+            .into_py(py),
+        YrValue::Structure(s) => s
+            .iter()
+            .map(|(k, v)| (k, yr_value_to_py_object(v)))
+            .collect::<HashMap<_, _>>()
+            .into_py(py),
+    }
+}
+
+#[pymethods]
+impl Context {
+    /// Context constructor.
+    ///
+    /// Call without arguments will initialize default context matching /dev/null, without rule or
+    /// any module data. `sample` and `rule` are expected to be valid paths. `module_data` is a
```

Comment:
nitpick: rule is not a path, but the ruleset text itself
I'm testing rule construction using the rule text and it doesn't seem to compile, even when i fill the `rule` argument with ruleset from a unit test that has been working previously. Not even when i fill it with a simple `rule abc { condition: 1 }` rule. I suspect it could be the result of some special characters being misinterpreted during the conversion from python string to internal representation, but it's only a guess.

Reply:
Changed the rule to more clearly reflect the usage.

Also added tests for Contexts with rules.